### PR TITLE
Fix wildcard rules

### PR DESCRIPTION
**Pull request by oscarbc96**
### Improvements
- Refactored the `KMSKeyWildcardPrincipalRule` rule
### Updates
- Update `GenericWildcardPrincipalRule`, `PartialWildcardPrincipalRule` and `GenericResourcePartialWildcardPrincipalRule` message
- Update docs
### Fixes
- Fix `GenericWildcardPrincipalRule` that could add a false-positive
- Fix `GenericWildcardPrincipalRule` that wasn't handling canonical IDs
- Fix unit tests

### CODE REVIEW

File: cfripper/rules/wildcard_principals.py [hunk 1]
```diff
@@ -104,15 +102,16 @@ def check_for_wildcards(
         extras: Optional[Dict] = None,
     ):
         for statement in resource.statement_as_list():
-            if statement.Effect == "Allow" and statement.principals_with(self.FULL_REGEX):
-                for principal in statement.get_principal_list():
-                    account_id_match = self.IAM_PATTERN.match(principal) or self.AWS_ACCOUNT_ID_PATTERN.match(principal)
-                    account_id = account_id_match.group(1) if account_id_match else None
+            filtered_principals = statement.principals_with(self.FULL_REGEX)
+            if statement.Effect == "Allow" and filtered_principals:
+                for principal in filtered_principals:
```
Comment (oscarbc96) [inline cfripper/rules/wildcard_principals.py:107]: Previously we iterated over all principals, but we should iterate just over the ones that match the principal

File: cfripper/rules/wildcard_principals.py [hunk 2]
```diff
@@ -104,15 +102,16 @@ def check_for_wildcards(
...
+                    # if we can't find the account ID it might be a canonical ID
+                    identifier = get_account_id_from_principal(principal) or principal
```
Comment (oscarbc96) [inline cfripper/rules/wildcard_principals.py:109]: Canonical IDs weren't validated and could rase a False-positive

File: tests/config/test_config.py [hunk 3]
```diff
@@ -92,26 +92,6 @@ def test_load_filters_work_with_several_rules(template_two_roles_dict, test_file
...
-            Failure(
-                granularity=RuleGranularity.RESOURCE,
-                reason="RootRoleTwo should not allow wildcard in principals or account-wide principals (principal: 'arn:aws:iam::123456789:user/someuser@bla.com')",
-                risk_value=RuleRisk.MEDIUM,
-                rule="PartialWildcardPrincipalRule",
-                rule_mode=RuleMode.BLOCKING,
-                actions=None,
-                resource_ids={"RootRoleTwo"},
-                resource_types={"AWS::IAM::Role"},
-            ),
```
Comment (oscarbc96) [inline tests/config/test_config.py:114]: This test was wrong as a result of iterating all principals and not some of them in GenericWildcardPrincipalRule. So we were adding false-positive.

**Review by w0rmr1d3r — APPROVED** [review]
Really nice job!

**Review by ignaciobolonio — APPROVED** [review]

**Pull request merged**